src/ledger/LedgerManagerImpl.h

```diff
@@ -42,58 +42,64 @@ class BasicWork;
         std::shared_ptr<SorobanNetworkConfig const> sorobanConfig;
         HistoryArchiveState has;
         std::shared_ptr<SearchableLiveBucketListSnapshot const> snapshot;
     };
 
-    Application& mApp;
-    std::unique_ptr<XDROutputFileStream> mMetaStream;
-    std::unique_ptr<XDROutputFileStream> mMetaDebugStream;
-    std::weak_ptr<BasicWork> mFlushAndRotateMetaDebugWork;
-    std::filesystem::path mMetaDebugPath;
+    // Any state that apply needs to access through the app connector should go
+    // here, at very least just to make it clear what is being accessed by which
+    // threads. We may try to further encapsulate it.
+    struct ApplyState
+    {
+        // Latest Soroban config during apply (should not be used outside of
+        // application, as it may be in half-valid state). Note that access to
+        // this variable is not synchronized, since it should only be used by
+        // one thread (main or ledger close).
+        std::shared_ptr<SorobanNetworkConfig> mSorobanNetworkConfig;
+    };
 
-  private:
-    // Cache LCL state, accessible only from main thread
-    LedgerHeaderHistoryEntry mLastClosedLedger;
-
-    // Read-only Soroban network configuration, accessible by main thread only.
-    // This config has to match the read-only ledger state snapshot managed by
-    // LedgerManager. Today this is only accessed by the main thread, but it can
-    // be used by multiple threads as long snapshot consistency
-    // requirement is preserved (though synchronization should be added in this
-    // case)
-    std::shared_ptr<SorobanNetworkConfig const> mSorobanNetworkConfigReadOnly;
-
-    // Latest Soroban config during apply (should not be used outside of
-    // application, as it may be in half-valid state). Note that access to this
-    // variable is not synchronized, since it should only be used by one thread
-    // (main or ledger close).
-    std::shared_ptr<SorobanNetworkConfig> mSorobanNetworkConfigForApply;
-
-    // Cache most recent HAS, accessible only from main thread
-    HistoryArchiveState mLastClosedLedgerHAS;
+    struct LedgerMetrics
+    {
+        medida::Timer& mTransactionApply;
+        medida::Histogram& mTransactionCount;
+        medida::Histogram& mOperationCount;
+        medida::Histogram& mPrefetchHitRate;
+        medida::Timer& mLedgerClose;
+        medida::Buckets& mLedgerAgeClosed;
+        medida::Counter& mLedgerAge;
+        medida::Counter& mTransactionApplySucceeded;
+        medida::Counter& mTransactionApplyFailed;
+        medida::Counter& mSorobanTransactionApplySucceeded;
+        medida::Counter& mSorobanTransactionApplyFailed;
+        medida::Meter& mMetaStreamBytes;
+        medida::Timer& mMetaStreamWriteTime;
+        LedgerMetrics(medida::MetricsRegistry& registry);
+    };
+
+    // This state is private to the apply thread and holds work-in-progress
+    // that gets accessed via the AppConnector, from inside transactions.
+    ApplyState mApplyState;
```

Comment:
Might be nice to make access to this a `getter` wrapper function that asserts threadIsApply. That might be too much work now though because we'd have to pass in the CFG to see if background apply is enabled or not, but would be nice to see in the future.

Reply:
I wanted to, but apply can happen on main thread of apply thread, even when background apply is enabled. This is still fairly messy and we might want to prohibit both possibilities, but as of today it's how it is.